### PR TITLE
 added 2, 22, 692, 994, 146

### PR DESCRIPTION
**Pull request by gaizzzzi**

### CODE REVIEW

File: main/146_LRU_Cache.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        self.prev = _prev
+        self.next = _next
+        
+class LRUCache:
+    # 21:35 - 22:00
+    # hash + double linked list
+    def __init__(self, capacity: int):
+        self.capacity = capacity
+        self.map = {}
+        self.size = 0
```
Comment (sheep-sheep) [inline main/146_LRU_Cache.py:14]: `size` can be `len(self.map)`

File: main/146_LRU_Cache.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            
+        else:
+            if self.size == self.capacity:
+                self._delete()
+            else:
+                self.size += 1
+            self.map[key] = node(key, value)
+        self._update_tail(key)
+    
+    def _remove_current(self, key):
```
Comment (sheep-sheep) [inline main/146_LRU_Cache.py:41]: nit: can rename to `_remove_current_node` or `_update_current_node`

File: main/146_LRU_Cache.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    
+    def _update_tail(self, key):
+        self.map[key].prev = self.tail.prev
+        self.tail.prev.next = self.map[key]
+        self.tail.prev = self.map[key]
+        self.map[key].next = self.tail
+    
+    def _delete(self):
+        _map = self.head.next
+        del self.map[_map.key]
```
Comment (sheep-sheep) [inline main/146_LRU_Cache.py:53]: * `_map` is redundant and the naming is weird.
* can rename method name to `_delete_lru_node`

File: main/22_Generate_Parentheses.py [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+        ans = []
+        def helper(left, right, path):
+            if not left and not right:
+                ans.append(path)
+            if left > 0:
+                helper(left - 1, right, path + "(")
+            if left < right:
+                helper(left, right - 1, path + ")")
+        helper(n, n, "")
+        return ans
```
Comment (sheep-sheep) [inline main/22_Generate_Parentheses.py:13]: :six::six::six::six::six::six::six::six::six::six::six:

File: main/692_Top_K_Frequent_Words.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+        hp = []
+        word_map = {}
+        for word in words:
+            word_map[word] = word_map.get(word, 0) + 1
+        for word, count in word_map.items():
+            if len(hp) == k:
+                heappushpop(hp, (count, comparison(word)))
+            else:
+                heappush(hp, (count, comparison(word)))
+        ans = [heappop(hp)[1].word for _ in range(k)]
```
Comment (sheep-sheep) [inline main/692_Top_K_Frequent_Words.py:21]: What's the size of `hp`? Isn't it larger than `k`?

File: main/7_Reverse_Integer.py [hunk 6]
```diff
@@ -0,0 +1,15 @@
+class Solution:
+    def reverse(self, x: int) -> int:
+        # 19:50 - 20:00
+        maxint = str(1 << 32 - 1)
+        minint = str(-(1 << 31))
```
Comment (sheep-sheep) [inline main/7_Reverse_Integer.py:5]: Python doesn't have max and min int  i think

File: main/7_Reverse_Integer.py [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+        str_x = str(x)
+        if str_x[0] == "-":
```
Comment (sheep-sheep) [inline main/7_Reverse_Integer.py:7]: can be simplified like: return `int('-'+str(abs(x))[::-1] if x<0 else str(abs(x))[::-1])`

Comment (sheep-sheep) [reply main/7_Reverse_Integer.py:7]:
```
class Solution(object):
    def reverse(self, x):
        """
        :type x: int
        :rtype: int
        """
        sign = False
        if x < 0:
            sign = True
        x = abs(x)
        res = 0
        while x:
            # print x, res
            res = res*10 + x%10
            # print res
            x = x/10
        if res > 2**31:
            return 0
        if sign:
            return -res
        return res
```

File: main/146_LRU_Cache.py (hunk 2 above)
Comment (gaizzzzi) [reply main/146_LRU_Cache.py:41]: > nit: can rename to `_remove_current_node` or `_update_current_node`

nice

File: main/146_LRU_Cache.py (hunk 3 above)
Comment (gaizzzzi) [reply main/146_LRU_Cache.py:53]: > * `_map` is redundant and the naming is weird.
> * can rename method name to `_delete_lru_node`

🐒

File: main/692_Top_K_Frequent_Words.py (hunk 5 above)
Comment (gaizzzzi) [reply main/692_Top_K_Frequent_Words.py:21]: > What's the size of `hp`? Isn't it larger than `k`?

hp will always stay size no greater than k

**Pull request merged**